src/libraries/System.Collections/src/System/Collections/Generic/PriorityQueue.cs

```diff
@@ -0,0 +1,695 @@
+            var node = (element, priority);
+            _nodes[_size] = node;
+            _size++;
+            _version++;
+
+            // Restore the heap order
+            int lastNodeIndex = GetLastNodeIndex();
+            MoveUp(node, lastNodeIndex);
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Peek()
+        {
+            if (TryPeek(out TElement? element, out TPriority? priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">The queue is empty.</exception>
+        public TElement Dequeue()
+        {
+            if (TryDequeue(out TElement? element, out TPriority? priority))
+            {
+                return element;
+            }
+            else
+            {
+                throw new InvalidOperationException(SR.InvalidOperation_EmptyQueue);
+            }
+        }
+
+        /// <summary>
+        /// Dequeues the element associated with the minimal priority
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryDequeue([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (_size != 0)
+            {
+                (element, priority) = _nodes[RootIndex];
+                Remove(RootIndex);
+                return true;
+            }
+
+            element = default;
+            priority = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the element associated with the minimal priority.
+        /// </summary>
+        /// <returns>
+        /// <see langword="true"/> if the priority queue is non-empty; <see langword="false"/> otherwise.
+        /// </returns>
+        public bool TryPeek([MaybeNullWhen(false)] out TElement element, [MaybeNullWhen(false)] out TPriority priority)
+        {
+            if (_size != 0)
+            {
+                (element, priority) = _nodes[RootIndex];
+                return true;
+            }
+
+            element = default;
+            priority = default;
+            return false;
+        }
+
+        /// <summary>
+        /// Combined enqueue/dequeue operation, generally more efficient than sequential Enqueue/Dequeue calls.
+        /// </summary>
+        public TElement EnqueueDequeue(TElement element, TPriority priority)
+        {
+            if (element is null)
```

Comment:
Why would we need to check if either `element` or `priority` are null? Like `Queue<string?>` it should be possible to enqueue null elements.

Reply:
Yup, such null checks should be removed.  If we actually want to prohibit null priorities, TPriority should be defined with a notnull constraint, but I suspect we don't want such a limitation.